File: src/code-samples/langchain/long-term-memory-storage-postgres.py
# :snippet-start: long-term-memory-storage-postgres-py
from collections.abc import Sequence

from langgraph.store.base import IndexConfig
from langgraph.store.postgres import PostgresStore  # type: ignore[import-not-found]


def embed(texts: Sequence[str]) -> list[list[float]]:
    # Replace with an actual embedding function or LangChain embeddings object
    return [[1.0, 2.0] for _ in texts]


DB_URI = "postgresql://postgres:postgres@localhost:5432/postgres?sslmode=disable"
# :remove-start:
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from conftest import get_postgres_uri, prepare_postgres_store

DB_URI = get_postgres_uri()
prepare_postgres_store(DB_URI)
# :remove-end:

with PostgresStore.from_conn_string(
    DB_URI,
    index=IndexConfig(embed=embed, dims=2),  # type: ignore[arg-type]
) as store:
    store.setup()
    user_id = "my-user"
    application_context = "chitchat"
    namespace = (user_id, application_context)
    store.put(
        namespace,
        "a-memory",
        {
            "rules": [
                "User likes short, direct language",
                "User only speaks English & python",
            ],
            "my-key": "my-value",
        },
    )
    item = store.get(namespace, "a-memory")
    items = store.search(
        namespace, filter={"my-key": "my-value"}, query="language preferences"
    )
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    assert item is not None
    assert item.value["my-key"] == "my-value"
    assert "rules" in item.value
    assert len(item.value["rules"]) == 2
    assert len(items) > 0
    print("✓ PostgresStore operations work correctly")
# :remove-end:

File: src/code-samples/langchain/long-term-memory-storage-postgres.ts
// :snippet-start: long-term-memory-storage-postgres-js
import { PostgresStore } from "@langchain/langgraph-checkpoint-postgres/store";

const embed = (texts: string[]): number[][] => {
  return texts.map(() => [1.0, 2.0]);
};

const DB_URI =
  process.env.POSTGRES_URI ??
  "postgresql://postgres:postgres@localhost:5432/postgres?sslmode=disable";
const store = PostgresStore.fromConnString(DB_URI, {
  index: { embed, dims: 2 },
});
// :remove-start:
// Drop tables from prior runs (Python samples use different schema; CI shares one DB)
await (
  store as { core: { pool: { query: (q: string) => Promise<unknown> } } }
).core.pool.query(
  "DROP TABLE IF EXISTS public.store_vectors CASCADE; DROP TABLE IF EXISTS public.store CASCADE; DROP TABLE IF EXISTS public.store_migrations CASCADE;",
);
// :remove-end:
await store.setup();

const userId = "my-user";
const applicationContext = "chitchat";
const namespace = [userId, applicationContext];

await store.put(namespace, "a-memory", {
  rules: [
    "User likes short, direct language",
    "User only speaks English & TypeScript",
  ],
  "my-key": "my-value",
});

const item = await store.get(namespace, "a-memory");
const items = await store.search(namespace, {
  filter: { "my-key": "my-value" },
  query: "language preferences",
});
// :snippet-end:

// :remove-start:

try {
  await store.setup();

  const userId = "my-user";
  const applicationContext = "chitchat";
  const namespace = [userId, applicationContext];

  await store.put(namespace, "a-memory", {
    rules: [
      "User likes short, direct language",
      "User only speaks English & TypeScript",
    ],
    "my-key": "my-value",
  });

  const item = await store.get(namespace, "a-memory");
  const items = await store.search(namespace, {
    filter: { "my-key": "my-value" },
    query: "language preferences",
  });

  // Verify the operations work
  if (!item) {
    throw new Error("Item should not be null");
  }
  if (item.value["my-key"] !== "my-value") {
    throw new Error('Expected my-key to be "my-value"');
  }
  if (!item.value["rules"]) {
    throw new Error("Expected rules to exist");
  }
  if ((item.value["rules"] as string[]).length !== 2) {
    throw new Error("Expected 2 rules");
  }

  // Verify search returns results
  if (items.length === 0) {
    throw new Error("Expected search to return results");
  }

  console.log("✓ PostgresStore operations work correctly");
} finally {
  await store.stop();
}

// :remove-end:
